src/boot/efi/cpio.c

```diff
@@ -0,0 +1,465 @@
+        a = (CHAR8*) *cpio_buffer + *cpio_buffer_size;
+
+        CopyMem(a, "070701", 6); /* magic ID */
+        a += 6;
+
+        a = write_cpio_word(a, (*inode_counter)++);                         /* inode */
+        a = write_cpio_word(a, access_mode | 0100000 /* = S_IFREG */);      /* mode */
+        a = write_cpio_word(a, 0);                                          /* uid */
+        a = write_cpio_word(a, 0);                                          /* gid */
+        a = write_cpio_word(a, 1);                                          /* nlink */
+
+        /* Note: we don't make any attempt to propagate the mtime here, for two reasons: it's a mess given
+         * that FAT usually is assumed to operate with timezoned timestamps, while UNIX does not. More
+         * importantly though: the modifications times would hamper our goals of providing stable
+         * measurements for the same boots. After all we extend the initrds we generate here into TPM2
+         * PCRs. */
+        a = write_cpio_word(a, 0);                                          /* mtime */
+        a = write_cpio_word(a, contents_size);                              /* size */
+        a = write_cpio_word(a, 0);                                          /* major(dev) */
+        a = write_cpio_word(a, 0);                                          /* minor(dev) */
+        a = write_cpio_word(a, 0);                                          /* major(rdev) */
+        a = write_cpio_word(a, 0);                                          /* minor(rdev) */
+        a = write_cpio_word(a, target_dir_prefix_size + fname_size + 1);    /* fname size */
+        a = write_cpio_word(a, 0);                                          /* "crc" */
+
+        CopyMem(a, target_dir_prefix, target_dir_prefix_size);
+        a += target_dir_prefix_size;
+        *(a++) = '/';
+        a = mangle_filename(a, fname);
+
+        /* Pad to next multiple of 4 */
+        a = pad4(a, *cpio_buffer);
+
+        CopyMem(a, contents, contents_size);
+        a += contents_size;
+
+        /* Pad to next multiple of 4 */
+        a = pad4(a, *cpio_buffer);
+
+        assert(a == (CHAR8*) *cpio_buffer + *cpio_buffer_size + l);
+        *cpio_buffer_size += l;
+
+        return EFI_SUCCESS;
+}
+
+static EFI_STATUS pack_cpio_dir(
+                const CHAR8 *path,
+                UINT32 access_mode,
+                UINT32 *inode_counter,
+                VOID **cpio_buffer,
+                UINTN *cpio_buffer_size) {
+
+        UINTN l, path_size;
+        CHAR8 *a;
+
+        assert(path);
+        assert(cpio_buffer);
+        assert(cpio_buffer_size);
+
+        /* Serializes one directory inode in cpio format. Note that cpio archives must first create the dirs
+         * they want to place files in. */
+
+        if (*inode_counter == UINT32_MAX)
+                return EFI_OUT_OF_RESOURCES;
+
+        l = 6 + 13*8 + 1; /* Fixed CPIO header size, and NUL byte after the file name*/
```

Comment:
Maybe make a #define for these?

Reply:
dunno, if we use this exactly once then i think it's easier to grok without indirection via a define. Moreover macros "leak" into the rest of the source file (unless `#undef`fed again, but that's brittle) so I think it's not the right choice.